Database-design

K-Hir-pk18 opened the pull request:
Readmeの作成
ER図（furima-30316/draw.dio）の作成

K-Hir-pk18 committed:
Database-design

Readmeの作成
ER図（furima-30316/draw.dio）の作成

Katsuno0814 commented:
レビュー依頼ありがとうございます。
プルリクエストにwhat,whyの記載がございません。
What（どのような実装をしているのか）と Why（なぜこの実装が必要なのか）を記載することで、第三者にブランチでの作業内容を伝えることが可能でございます。
その後、再度レビュー依頼フォームにてご提出をお願いいたします。

K-Hir-pk18 committed:
database-design2

##what
(ER図)
・furima-30316/draw.dio

(readme)
・README.md

##why
アプリ開発を円滑に進める為、
データベースを最初に設計した。

K-Hir-pk18 commented:
> レビュー依頼ありがとうございます。
> プルリクエストにwhat,whyの記載がございません。
> What（どのような実装をしているのか）と Why（なぜこの実装が必要なのか）を記載することで、第三者にブランチでの作業内容を伝えることが可能でございます。
> その後、再度レビュー依頼フォームにてご提出をお願いいたします。

ご対応ありがとうございます。
再度お願いいたします。

iy06 commented on README.md:
```diff
@@ -13,6 +13,74 @@ Things you may want to cover:
+
+### Association
+- has_many :Products
+- has_many :Purchases
```
Usersテーブル

- mail_address こちらはdeviseのデフォルトであるemailに修正しましょう。
- passwordこちらもdeviseのデフォルトであるencrypted_passwordに修正しましょう。
- birth_year, month, dateこちらはdate型で定義し、一つのカラムにまとめましょう。

iy06 commented on README.md:
```diff
@@ -13,6 +13,74 @@ Things you may want to cover:
+### Association
+-belongs_to :Users
+-has_one_attached :Images
+-has_one_attached :Purchases
```
Productsテーブル

- category, condition, send_charge, send_place, send_spanこちらはactive_hashを使用しますのでカラム名に_idをつけましょう。

iy06 commented on README.md:
```diff
@@ -13,6 +13,74 @@ Things you may want to cover:
+| product    | reference | foreign_key: true |
+
+### Association
+-belongs_to :Products
```
Imagesテーブル

こちらは、active_storageによって自動でマイグレーションファイルが作成されるため不要です。
もし記載するのであれば、実際に生成された段階でマイグレーションファイルをもとに作成しましょう。

iy06 commented on README.md:
```diff
@@ -13,6 +13,74 @@ Things you may want to cover:
+
+### Association
+-belongs_to :Users
+-belongs_to :Products
```
Purchasesテーブル

- prefectureこちらは、Productsテーブルのsend_placeと同じactive_hashを使用しますのでカラム名をどちらかに統一しましょう。
- telephoneこちらは、string型で定義しましょう。0から始まる数字を保存するためです。

K-Hir-pk18 replied:
iy06様

細やかな点までご対応ありがとうございます。

active_hashだと末尾に_idがつくのですね！
まったく意識していませんでした。
勉強になります。

iy06 reviewed (commented):
レビュー依頼ありがとうございます！

修正箇所をコメントさせて頂きましたので、ご確認お願い致します。

K-Hir-pk18 committed:
Database Design3

##what
(ER図)draw.dio

　・Users table
　　→　カラム名の変更
　　　　　email,encrypted_password
　　→　birth〜カラムの統合
　　→　birth_profileのdate型化

　・Products table
　　→　send_palaceの名称変更
　　　　　prefectureへ
　　→　active_hash踏まえ_idの追加
　　　　　category, condition, 　　　　　send_charge, prefecture,
　　　　　send_span

　・image table
　　→　削除
　　　（自動マイグレーションのため）

　・Purchases table
　　→　active_hash踏まえ_idの追加
　　　　　prefecture
　　→　telephoneのstring型化

(README)README.md
　　→上記踏まえた内容に更新

##why
　pullリクエストの指摘点対応

(追加質問)
usersテーブルの
familiname_kanji とfirstname_kanji
familiname_kana とfirstname_kana
これらも一緒にできますか？

yuevo commented on README.md:
```diff
@@ -13,6 +13,56 @@ Things you may want to cover:
+## Users
+|Column              |Type     |Options                    |
+|--------------------|---------|---------------------------|
+| nickname           | string  | null: false, unique: true |
```
```
| nickname           | string  | null: false, unique: true |
```
今回のusersテーブルで一意性のみを許可したいのは`email`カラムのみになりますので
こちらの`unique: true `は削除しておきましょう！

yuevo commented on README.md:
```diff
@@ -13,6 +13,56 @@ Things you may want to cover:
+
+### Association
+-belongs_to :Users
+-has_one_attached :Purchases
```
```
has_one_attached :Purchases
```
`has_one_attached`はまた別の意味になってしまうので、今回のアソシエーションでは`has_one`に修正しておきましょう！

yuevo commented on README.md:
```diff
@@ -13,6 +13,56 @@ Things you may want to cover:
+| user           | reference | foreign_key: true |
+
+### Association
+-belongs_to :Users
+-belongs_to :Products
```
```
-belongs_to :Users
-belongs_to :Products
```
`belongs_to`の場合、テーブル名は単数形になります。修正しておきましょう！

yuevo commented on README.md:
```diff
@@ -13,6 +13,56 @@ Things you may want to cover:
+| user_id        | reference | foreign_key: true |
+
+### Association
+-belongs_to :Users
```
```
-belongs_to :Users
```
`belongs_to`の場合、テーブル名は単数形になります。修正しておきましょう！

yuevo commented on README.md:
```diff
@@ -13,6 +13,56 @@ Things you may want to cover:
+| municipality   | string    | null: false       |
+| banch          | string    | null: false       |
+| building_name  | string    |                   |
+| telephone      | string    | foreign_key: true |
```
```
| telephone      | string    | foreign_key: true |
```
こちらは外部キーではないかと見受けられますので、`Options`の内容を修正しておきましょう！

yuevo commented on README.md:
```diff
@@ -13,6 +13,56 @@ Things you may want to cover:
+| banch          | string    | null: false       |
+| building_name  | string    |                   |
+| telephone      | string    | foreign_key: true |
+| product        | reference | foreign_key: true |
+| user           | reference | foreign_key: true |
```
この２つの外部キーはまた別のテーブルで管理するほうが好ましいです。
理由は、エンティティが異なるためです。
下記のようなテーブルに修正してみましょう！

```
【購入履歴テーブル】
購入者（user）
購入された商品（item）

【住所情報テーブル】
郵便番号〜電話番号
購入履歴との外部キー
```

K-Hir-pk18 replied:
yuevo様

チェックありがとうございます。
なかなか難しいですね…💦

K-Hir-pk18 replied:
ER図

https://gyazo.com/40c7206ecbf581725ffc3a41d8d4def5

yuevo reviewed (commented):
レビューのご依頼ありがとうございます！
６点コメントさせていただいました。
ご確認のほど、よろしくお願いいたします！

K-Hir-pk18 committed:
database_design4

##what
以下の修正

　・nickname option のunieque削除
　・has_one_attached　→　has_one
　・belongs_toとhas_oneの　単数化
　・purchase_logsテーブルの設定

##why
コードレビューの対応

shodai-ito commented on README.md:
```diff
@@ -13,6 +13,66 @@ Things you may want to cover:
+| send_charge_id | integer   | null: false       |
+| prefecture_id  | integer   | null: false       |
+| send_span_id   | integer   | null: false       |
+| user_id        | reference | foreign_key: true |
```
user_idの_idを削除しておきましょう！
references型の場合、カラム名に自動で_idが付与されます。
現在の記述ですと、DB作成時にuser_id_idとなってしまうためです。

shodai-ito commented on README.md:
```diff
@@ -13,6 +13,66 @@ Things you may want to cover:
+
+### Association
+-belongs_to :User
+-belongs_to :Product
```
Purchaseテーブルとのアソシエーションを記述しましょう！

shodai-ito commented on README.md:
```diff
@@ -13,6 +13,66 @@ Things you may want to cover:
+| telephone      | string    | null: false       |
+
+### Association
+-has_one :Purchase_log
```
こちらアソシエーションをbelongs_toに変更されてはいかがでしょうか？
Purchase_logテーブルとPurchaseテーブルの関係性を考えた場合、「一回の購入につき、一つの住所情報が必要」となる方が自然かと思われるためです。

shodai-ito commented on README.md:
```diff
@@ -13,6 +13,66 @@ Things you may want to cover:
+| municipality   | string    | null: false       |
+| banch          | string    | null: false       |
+| building_name  | string    |                   |
+| telephone      | string    | null: false       |
```
Purchase_logのidを保存する外部キーカラムを追加しましょう！

K-Hir-pk18 replied:
shodai-ito様

チェック頂きありがとうございます！
自分だと細かな点まで気が回らないですね💦

お名前気になったのですが、
今は３代目ぐらいまでいるのでしょうか？？

K-Hir-pk18 replied:
ER図

https://gyazo.com/35ed78ef12d981ae60b5aae9678a9c2d

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。
DB設計完了まで後僅かです！

K-Hir-pk18 committed:
database design5

##what
productsテーブルのuser_id修正

purchase_logsアソシエーションに
　has_one :purchase追加

purchaseテーブルのpurchase_logs
　テーブルのassociation変更
　has_one　→　belongs_to

purchasesテーブルにpurchase_logカラム追加

##why
コードレビューの対応

yogotaku commented:
レビュー依頼ありがとうございます。
修正確認いたしました。LGTMとさせていただきます！
ブランチをマージした後、次の実装へ進んでください。
お疲れ様でした！引き続き実装頑張っていきましょう！

K-Hir-pk18 commented:
yogataku様

ありがとうございます！
オーバーかもしれませんが、多くのメンターさんに助けてもらいつつも
達成できてとても嬉しいです！！
